Создали новый класс FileBackedTaskManager отвечающий за запись задач …

Mofo328 opened the pull request:
…в файлы, и наоборот

В каждый класс задач,добавили типы задач
Добавили создание обьекта нового класса в менеджер

Mofo328 committed:
Создали новый класс FileBackedTaskManager отвечающий за запись задач в файлы, и наоборот
В каждый класс задач,добавили типы задач
Добавили создание обьекта нового класса в менеджер

Mofo328 committed:
исправил логику в loadFromFile

Mofo328 committed:
исправил логику в loadFromFile

Ivanov-s-tmn commented on src/model/Task.java:
```diff
@@ -7,18 +7,25 @@ public class Task {
     protected String name;
     protected Status status;
     protected String description;
+    protected TypeOfTask type;
```
Можно не добавлять поле, достаточно просто сделать геттеры, которые будут возвращать тип в зависимости от класса, в котором геттер реализован

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,187 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
Реализовывать интерфейс `TaskManager` не обязательно, он уже реализуется `InMemoryTaskManager`

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,187 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
+
+
```
Лишний отступ

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,187 @@
+        save();
+    }
+
+    @Override
+    public Task get(int id) throws ManagerSaveException {
+        Task taskByIdFromFile = super.get(id);
+        save();
+        return taskByIdFromFile;
+    }
```
Методы гет никак не будут влиять на файл, поэтому их не нужно переопределять с логикой сохранения в файл. Мы не сохраняем историю

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,187 @@
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
+
+
+    static File file;
```
Статик поле лучше не делать, это делает его общим для всех файловых менеджеров, а их может быть несколько. Плюс, не забывай про модификаторы доступа

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,187 @@
+            for (Task task : allTasks) {
+                fileWriter.write(toString(task) + "\n");
+            }
+
+
```
Лишние отступы

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,187 @@
+            return new SubTask(name, status, description, getEpicId);
+        }
+        return null;
+    }
```
Методы преобразования из строки/в строку лучше будет вынести в отдельный утилитный класс, чтобы разгрузить менеджер. 
В методе `fromString()` лучше заменить if else на switch case, он хорошо подходит для работы с енамами, плюс switch case доступный с 12 версии языка займет меньше места в коде)

Ivanov-s-tmn commented on src/exceptions/ManagerSaveException.java:
```diff
@@ -0,0 +1,7 @@
+package exceptions;
+
+public class ManagerSaveException extends Exception{
```
Унаследуй свое исключение от `RuntimeException` - непроверяемого типа, тогда его не нужно будет проверять и прокидывать из методов

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,187 @@
+                } else if (task.getType() == TypeOfTask.EPIC) {
+                    fileBackedTaskManager.createEpic((Epic) task);
+                } else if (task.getType() == TypeOfTask.SUBTASK) {
+                    fileBackedTaskManager.createSubTask((SubTask) task);
```
вызов методов create будет перезаписывать данные в файл, добавляй задачи напрямую в хешмапу

Ivanov-s-tmn commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,56 @@
+        assertEquals("0,Имя,NEW,Описание,TASK,", fileBackedTaskManager.toString(task), "Incorrect toString output for Task");
+        assertEquals("0,Имя,NEW,Описание,EPIC,", fileBackedTaskManager.toString(epic), "Incorrect toString output for Task");
+        assertEquals("0,Имя,NEW,Описание,SUBTASK,2", fileBackedTaskManager.toString(subTask), "Incorrect toString output for Task");
+    }
```
Создай файловый менеджер, заполни его задачами, потом создай файловый менеджер из того же файла через `loadFromFile()` и проверь, что списки задач/подзадач/эпиков этих менеджеров одинаковы (второй менеджер правильно восстановил свое состояние из файла)

Mofo328 committed:
вынес логику преобразования строк в отдельный класс Converter.
Во всех классах задач,убрал поле Type,возвращаю тип через гетер
FileBackedTaskManager в классе, убрал не нужные метода получения задач,изменил поле Fail на со статик на приват
в методе loadFoam вместа записи задач,через методы create,кладу их сразу в мапу

Mofo328 committed:
вынес логику преобразования строк в отдельный класс Converter.
Во всех классах задач,убрал поле Type,возвращаю тип через гетер
FileBackedTaskManager в классе, убрал не нужные метода получения задач,изменил поле Fail на со статик на приват
в методе loadFoam вместа записи задач,через методы create,кладу их сразу в мапу

Ivanov-s-tmn commented on src/service/Converter.java:
```diff
@@ -0,0 +1,31 @@
+        Status status = Status.valueOf(values[1]);
+        String description = values[2];
+        TypeOfTask type = TypeOfTask.valueOf(values[3]);
+        int getEpicId = Integer.parseInt(values[4]);
```
Это поле нужно парсить только для сабтасок, мы упадем с исключением при попытке распарсить null в Integer

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -8,122 +8,100 @@
+    private final File file;
 
-    static File file;
+    private Converter converter;
```
Сделай метов `toString()` конвертера статическим и обращайся к нему через класс, сейчас `converter` null и будет NPE на строке `converter.toString(task)`. При статических методах, поле конвертера будет не нужно и не нужно будет его инициализировать

Ivanov-s-tmn commented on src/service/FileBackedTaskManager.java:
```diff
@@ -133,50 +111,20 @@ public void save() throws ManagerSaveException {
+                switch (task.getType()) {
+                    case TASK -> fileBackedTaskManager.tasks.put(task.getId(), task);
+                    case EPIC -> fileBackedTaskManager.epics.put(task.getId(), (Epic) task);
+                    case SUBTASK -> fileBackedTaskManager.subTasks.put(task.getId(), (SubTask) task);
```
Сабтаски нужно не забыть добавить в список внутри эпика

Ivanov-s-tmn commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -1,56 +1,30 @@
+        fileBackedTaskManager.createEpic(epic1);
+        fileBackedTaskManager.createSubTask(subTask1);
+        File saveFile = new File("tasks.txt");
+        FileBackedTaskManager fileBackedTaskManager2 = Managers.getDefaultFileBackedTaskManager();
+        FileBackedTaskManager.loadFromFile(saveFile);
```
Инициализируй второй менеджер через `loadFromFile()`

Mofo328 committed:
Перегрузил конструкторы в классах задач,чтобы задачи корректно создавались из файла
Упростил логигу toString класса Converter
В классе FileBackedTaskManager в метод loadFromFile добавил логику сохранения подзадач в список епиков,добавил логику реализации максимального айди

Ivanov-s-tmn commented on src/service/Converter.java:
```diff
@@ -4,28 +4,25 @@
-            case SUBTASK -> new SubTask(name, status, description, getEpicId);
-        };
+        String[] values = value.split("\n");
+        for (String s : values) {
```
А зачем тут нужен цикл? Ты же каждую строк по отдельности парсишь

Ivanov-s-tmn commented on src/service/Managers.java:
```diff
@@ -3,12 +3,16 @@
 
 public class Managers {
-    static File file = new File("file.csv");
+    static File file = new File("C:\\Users\\Win\\IdeaProjects\\java-kanban\\resources\\tasks.csv");
```
Не стоит тут делать статик поле для файла, можно просто в методе `getDefaultFileBackedTaskManager()` возвращать  экземпляр менеджера и передавать базовый путь к файлу, как пример
`return new FileBackedTaskManager(new File("resources/tasks.csv"));`
В тестах уже будет лучше использовать другой файл для прогона именно тестов, к примеру создать файл по пути test/resources/tasks.csv и проверять восстановление состояния из него. Файл перед каждым тестом лучше создавать и удалять в методах BeforeEach и AfterEach, чтобы при расширении тестов они не зависели от данных прошлых тестов

Ivanov-s-tmn commented on src/model/Epic.java:
```diff
@@ -4,10 +4,20 @@
-    private final List<Integer> subTasks = new ArrayList<>();
+    protected final List<Integer> subTasksFromEpic = new ArrayList<>();
+
+    @Override
+    public Integer getEpicId() {
+        return super.getEpicId();
+    }
```
Методы идут после конструкторов

Mofo328 committed:
Поменял логику исходя из рекомендаций к тз

Mofo328 committed:
Merge branch 'main' into sprint_7-solution-in-file-manager

# Conflicts:
#	src/model/Epic.java
#	src/service/InMemoryHistoryManager.java

Mofo328 committed:
Поменял логику исходя из рекомендаций к тз
